ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java

```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+        }
+        if (bestCandidateToCompact != null) {
+          List<String> segmentsToCompact = new ArrayList<>(validDataPerLogSegments.getSecond()
+              .subMap(bestCandidateToCompact.firstLogSegmentName, true, bestCandidateToCompact.lastLogSegmentName, true)
+              .keySet());
+          details = new CompactionDetails(validDataPerLogSegments.getFirst(), segmentsToCompact);
+        }
       }
     }
     return details;
   }
+
+  /**
+   * Calculates and returns the {@link CostBenefitInfo} for a given candidate
+   * @param firstLogSegmentName first log segment name of the candidate
+   * @param lastLogSegmentName last log segment name of the candidate
+   * @param validDataSizePerLogSegment Valid data size map
+   * @param segmentCapacity Segment Capacity of the {@link BlobStore}
+   * @param segmentHeaderSize Segment header size of the {@link BlobStore}
+   * @param errorMargin Error margin to factor in while calculating the cost benefit
+   * @return the {@link CostBenefitInfo} for the given candidate
+   */
+  private CostBenefitInfo getCostBenefitInfo(String firstLogSegmentName, String lastLogSegmentName,
+      NavigableMap<String, Long> validDataSizePerLogSegment, long segmentCapacity, long segmentHeaderSize,
+      long errorMargin) {
+    // cost = total valid data size (or data that needs to be copied over)
+    // benefit = no of segments reclaimed = (total source segments - total target segments)
+    //           total target segments = Ceil (total cost / (segment capacity - segment header - error margin))
+    long totalCost = 0;
+    int totalSegmentsToBeCompacted = 0;
+    String curSegmentName = firstLogSegmentName;
+    while (!curSegmentName.equals(lastLogSegmentName)) {
+      totalCost += validDataSizePerLogSegment.get(curSegmentName);
+      curSegmentName = validDataSizePerLogSegment.higherEntry(curSegmentName).getKey();
+      totalSegmentsToBeCompacted++;
+    }
+    // add the last log segment
+    totalCost += validDataSizePerLogSegment.get(curSegmentName);
+    totalSegmentsToBeCompacted++;
+    long maxCapacityPerSegment = segmentCapacity - segmentHeaderSize - errorMargin;
+    int benefit = totalSegmentsToBeCompacted - (int) Math.ceil(totalCost / (maxCapacityPerSegment * 1.0));
+    return new CostBenefitInfo(firstLogSegmentName, lastLogSegmentName, totalCost, benefit);
+  }
+
+  /**
+   * Stores cost benefit info for a candidate to compact
+   */
+  private class CostBenefitInfo {
+    long cost;
+    int benefit;
+    double costBenefitRatio;
+    String firstLogSegmentName;
+    String lastLogSegmentName;
+
+    CostBenefitInfo(String firstLogSegmentName, String lastLogSegmentName, long cost, int benefit) {
+      this.firstLogSegmentName = firstLogSegmentName;
+      this.lastLogSegmentName = lastLogSegmentName;
+      this.cost = cost;
+      this.benefit = benefit;
+      costBenefitRatio = benefit == 0 ? cost : cost / benefit;
```

Comment:
shouldn't this be `costBenefitRatio = benefit == 0 ? Double.MAX_VALUE : cost / benefit;`? Otherwise there would not be a difference b/w benefit = 0 and benefit = 1.

Also, you need to cast `cost` to `double` - otherwise the ratio you get is an integer (i.e 20/7 will result in 2.0).